GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs

```diff
@@ -0,0 +1,137 @@
+
+namespace GVFS.Service.UI
+{
+    public class GVFSToastRequestHandler
+    {
+        private const string GVFSAutomountStartTitle= "GVFS Automount";
+        private const string GVFSAutomountStartMessageFormat = "Attempting to mount {0} GVFS repos(s)";
+
+        private const string GVFSAutomountSuccessTitle = "GVFS Automount";
+        private const string GVFSAutomountSuccessMessageFormat = "The following GVFS repo is now mounted: {0}{1}";
+
+        private const string GVFSAutomountErrorTitle = "GVFS Automount";
+        private const string GVFSAutomountErrorMessageFormat = "The following GVFS repo failed to mount: {0}{1}";
+        private const string GVFSAutomountButtonTitle = "Retry";
+
+        private const string GVFSUpgradeTitleFormat = "New version {0} is available";
+        private const string GVFSUpgradeMessage = "Upgrade will unmount and remount gvfs repos, ensure you are at a stopping point. When ready, click Upgrade button to run upgrade.";
+        private const string GVFSUpgradeButtonTitle = "Upgrade";
+
+        private const string GvfsRemountActionPrefix = "gvfs mount ";
+        private const string GvfsUpgradeActionPrefix = "gvfs upgrade --confirm";
+
+        private ITracer tracer;
+        private IToastNotifier toastNotifier;
+
+        public GVFSToastRequestHandler(IToastNotifier toastNotifier, ITracer tracer)
+        {
+            this.toastNotifier = toastNotifier;
+            this.toastNotifier.UserResponseCallback = this.UserResponseCallback;
+            this.tracer = tracer;
+        }
+
+        public void HandleToastRequest(ITracer tracer, NamedPipeMessages.Notification.Request request)
+        {
+            string title = null;
+            string message = null;
+            string buttonTitle = null;
+            string args = null;
+
+            switch (request.Id)
+            {
+                case NamedPipeMessages.Notification.Request.Identifier.AutomountStart:
+                    {
+                        title = GVFSAutomountStartTitle;
+                        message = string.Format(GVFSAutomountStartMessageFormat, request.EnlistmentCount);
+                        break;
+                    }
+
+                case NamedPipeMessages.Notification.Request.Identifier.MountSuccess:
+                    {
+                        title = GVFSAutomountSuccessTitle;
+                        message = string.Format(GVFSAutomountSuccessMessageFormat, Environment.NewLine, request.Enlistment);
+                        break;
+                    }
+
+                case NamedPipeMessages.Notification.Request.Identifier.MountFailure:
+                    {
+                        title = GVFSAutomountErrorTitle;
+                        message = string.Format(GVFSAutomountErrorMessageFormat, Environment.NewLine, request.Enlistment);
+                        buttonTitle = GVFSAutomountButtonTitle;
+                        args = $"{GvfsRemountActionPrefix} {request.Enlistment}";

```

Comment:
Do we have to worry about escaping the enlistment here (spaces in the path) or is that done elsewhere?

Reply:
Added path validation both while sending and while receiving callbacks from toaster.